File: python/state.py
# ==============================================================
# GoldBot — state.py
# Bot runtime state: daily counters, setups, open positions,
# risk gates, and daily reset logic.
# ==============================================================

from __future__ import annotations
from dataclasses import dataclass, field

import config as cfg
import logger as log


# ══════════════════════════════════════════════════════════
# Data structures
# ══════════════════════════════════════════════════════════

@dataclass
class Setup:
    active:           bool  = False
    direction:        str   = ""
    created_time:     int   = 0    # epoch ms of the bar that formed the setup
    pullback_extreme: float = 0.0


@dataclass
class Position:
    mode:            str
    direction:       str
    size:            float
    entry:           float
    sl:              float
    tp1:             float
    tp2:             float
    tp1_done:        bool
    deal_id:         str
    deal_reference:  str
    opened_time:     int   # epoch ms


# ══════════════════════════════════════════════════════════
# Module-level state (singleton)
# ══════════════════════════════════════════════════════════

_day_realized_pnl_usd: float = 0.0
_trades_today:         int   = 0
_consecutive_losses:   int   = 0
_day_start_equity:     float = 0.0

_setup_scalp: Setup = Setup()
_setup_swing: Setup = Setup()

_open_positions: list[Position] = []


# ══════════════════════════════════════════════════════════
# Risk gates
# ══════════════════════════════════════════════════════════

def risk_ok() -> bool:
    if _trades_today >= cfg.MAX_TRADES_PER_DAY:
        log.warn(f"[Risk] Daily trade limit reached ({_trades_today}/{cfg.MAX_TRADES_PER_DAY})")
        return False
    if _day_realized_pnl_usd <= -cfg.DAILY_LOSS_LIMIT_USD:
        log.warn(f"[Risk] Daily loss limit reached (P&L: ${_day_realized_pnl_usd:.2f})")
        return False
    if _consecutive_losses >= cfg.MAX_CONSECUTIVE_LOSSES:
        log.warn(f"[Risk] Max consecutive losses reached ({_consecutive_losses})")
        return False
    return True


# ══════════════════════════════════════════════════════════
# Position tracking
# ══════════════════════════════════════════════════════════

def add_position(pos: Position) -> None:
    global _trades_today
    _open_positions.append(pos)
    _trades_today += 1
    log.info(f"[State] Position added: {pos.mode} {pos.direction} deal_id={pos.deal_id} | trades_today={_trades_today}")


def replace_position(old_deal_id: str, new_pos: Position) -> None:
    """Replace position without incrementing trades_today (used for TP1 reopen)."""
    global _open_positions
    _open_positions = [p for p in _open_positions if p.deal_id != old_deal_id]
    _open_positions.append(new_pos)
    log.info(f"[State] Position replaced: old={old_deal_id} → new={new_pos.deal_id}")


def remove_position(deal_id: str) -> None:
    global _open_positions
    _open_positions = [p for p in _open_positions if p.deal_id != deal_id]


def get_positions() -> list[Position]:
    return list(_open_positions)


# ══════════════════════════════════════════════════════════
# P&L tracking
# ══════════════════════════════════════════════════════════

def update_pnl(pnl_usd: float, is_loss: bool) -> None:
    global _day_realized_pnl_usd, _consecutive_losses
    _day_realized_pnl_usd += pnl_usd
    if is_loss:
        _consecutive_losses += 1
    else:
        _consecutive_losses = 0

    sign = "+" if pnl_usd >= 0 else ""
    log.info(
        f"[State] P&L update: {sign}${pnl_usd:.2f} | "
        f"day_pnl=${_day_realized_pnl_usd:.2f} | "
        f"consec_losses={_consecutive_losses}"
    )


# ══════════════════════════════════════════════════════════
# Setup accessors
# ══════════════════════════════════════════════════════════

def get_setup_scalp() -> Setup:  return _setup_scalp
def set_setup_scalp(s: Setup):
    global _setup_scalp
    _setup_scalp = s

def get_setup_swing() -> Setup:  return _setup_swing
def set_setup_swing(s: Setup):
    global _setup_swing
    _setup_swing = s


# ══════════════════════════════════════════════════════════
# Daily reset
# ══════════════════════════════════════════════════════════

def daily_reset(equity: float = 0.0) -> None:
    global _day_realized_pnl_usd, _trades_today, _consecutive_losses
    global _day_start_equity, _setup_scalp, _setup_swing
    _day_start_equity     = equity
    _day_realized_pnl_usd = 0.0
    _trades_today         = 0
    _consecutive_losses   = 0
    _setup_scalp          = Setup()
    _setup_swing          = Setup()
    log.info(f"[State] Daily reset. Start equity: ${equity:.2f}")


def get_stats() -> dict:
    return {
        "day_pnl":         _day_realized_pnl_usd,
        "trades_today":    _trades_today,
        "consec_losses":   _consecutive_losses,
        "open_count":      len(_open_positions),
    }

File: node/src/state.js
// ==============================================================
// GoldBot — state.js
// Bot runtime state: daily counters, setups, open positions,
// risk gates, and daily reset logic.
// ==============================================================

const cfg = require('./config');
const log = require('./logger');

// ── Daily counters ─────────────────────────────────────────
let dayRealizedPnlUsd = 0;
let tradesToday       = 0;
let consecutiveLosses = 0;
let dayStartEquity    = 0;

// ── Setup state ────────────────────────────────────────────
/** @type {{ active:boolean, direction?:string, createdTime?:number, pullbackExtreme?:number }} */
let setupScalp = { active: false };
/** @type {{ active:boolean, direction?:string, createdTime?:number, pullbackExtreme?:number }} */
let setupSwing = { active: false };

// ── Open positions (bot-tracked, complementing platform) ──
/**
 * @typedef {{ mode:string, direction:string, size:number,
 *             entry:number, sl:number, tp1:number, tp2:number,
 *             tp1Done:boolean, dealId:string,
 *             dealReference:string, openedTime:number }} Position
 * @type {Position[]}
 */
let openPositions = [];

// ══════════════════════════════════════════════════════════
// Risk gates
// ══════════════════════════════════════════════════════════

function riskOK() {
  if (tradesToday >= cfg.MAX_TRADES_PER_DAY) {
    log.warn(`[Risk] Daily trade limit reached (${tradesToday}/${cfg.MAX_TRADES_PER_DAY})`);
    return false;
  }
  if (dayRealizedPnlUsd <= -cfg.DAILY_LOSS_LIMIT_USD) {
    log.warn(`[Risk] Daily loss limit reached (P&L: $${dayRealizedPnlUsd.toFixed(2)})`);
    return false;
  }
  if (consecutiveLosses >= cfg.MAX_CONSECUTIVE_LOSSES) {
    log.warn(`[Risk] Max consecutive losses reached (${consecutiveLosses})`);
    return false;
  }
  return true;
}

// ══════════════════════════════════════════════════════════
// Position tracking
// ══════════════════════════════════════════════════════════

/**
 * Add a new position and increment tradesToday.
 * @param {Position} pos
 */
function addPosition(pos) {
  openPositions.push(pos);
  tradesToday += 1;
  log.info(`[State] Position added: ${pos.mode} ${pos.direction} dealId=${pos.dealId} | trades_today=${tradesToday}`);
}

/**
 * Adopt an existing platform position discovered at startup.
 * Does NOT increment tradesToday (the trade was already counted/not ours).
 * @param {Position} pos
 */
function adoptPosition(pos) {
  openPositions.push(pos);
  log.info(`[State] Position adopted from platform: ${pos.mode} ${pos.direction} dealId=${pos.dealId}`);
}

/**
 * Add a replacement position (TP1 partial-close reopen).
 * Does NOT increment tradesToday.
 */
function replacePosition(oldDealId, newPos) {
  openPositions = openPositions.filter(p => p.dealId !== oldDealId);
  openPositions.push(newPos);
  log.info(`[State] Position replaced: old=${oldDealId} → new=${newPos.dealId}`);
}

function removePosition(dealId) {
  openPositions = openPositions.filter(p => p.dealId !== dealId);
}

function getPositions() {
  return openPositions;
}

// ══════════════════════════════════════════════════════════
// P&L tracking
// ══════════════════════════════════════════════════════════

/**
 * Update day P&L and consecutive-loss counter after a close.
 * @param {number}  pnlUsd  Realised profit/loss in USD
 * @param {boolean} isLoss
 */
function updatePnL(pnlUsd, isLoss) {
  dayRealizedPnlUsd += pnlUsd;
  if (isLoss) {
    consecutiveLosses += 1;
  } else {
    consecutiveLosses = 0;
  }
  log.info(
    `[State] P&L update: ${pnlUsd >= 0 ? '+' : ''}$${pnlUsd.toFixed(2)} | ` +
    `day_pnl=$${dayRealizedPnlUsd.toFixed(2)} | consec_losses=${consecutiveLosses}`
  );
}

// ══════════════════════════════════════════════════════════
// Setup accessors
// ══════════════════════════════════════════════════════════

function getSetupScalp()   { return setupScalp; }
function setSetupScalp(s)  { setupScalp = s; }
function getSetupSwing()   { return setupSwing; }
function setSetupSwing(s)  { setupSwing = s; }

// ══════════════════════════════════════════════════════════
// Daily reset
// ══════════════════════════════════════════════════════════

/**
 * Reset all day-scoped counters and setups.
 * Called at UTC midnight and on startup.
 * @param {number} equity  Available account balance at reset time
 */
function dailyReset(equity = 0) {
  dayStartEquity    = equity;
  dayRealizedPnlUsd = 0;
  tradesToday       = 0;
  consecutiveLosses = 0;
  setupScalp        = { active: false };
  setupSwing        = { active: false };
  log.info(`[State] Daily reset. Start equity: $${dayStartEquity.toFixed(2)}`);
}

/**
 * Summary snapshot for status logging.
 */
function getStats() {
  return {
    dayRealizedPnlUsd,
    tradesToday,
    consecutiveLosses,
    openCount: openPositions.length,
  };
}

module.exports = {
  riskOK,
  addPosition, adoptPosition, replacePosition, removePosition, getPositions,
  updatePnL,
  getSetupScalp, setSetupScalp,
  getSetupSwing, setSetupSwing,
  dailyReset,
  getStats,
};
